plugins/ui-update/notifications.ts

```diff
@@ -0,0 +1,192 @@
+  _callbackType: ToastCallbackType,
+  toastNotification: ToastNotification
+) {
+  removeActiveNotification(toastNotification.id);
+}
+
+function setupUIUpdatedNotification(notificationService: NotificationService) {
+  getUiMetadata$().subscribe(uiMetadata => {
+    const { clientBuild, serverBuild } = uiMetadata;
+    const coercedClientBuild = semver.coerce(clientBuild || "");
+    const coercedServerBuild = semver.coerce(serverBuild || "");
+    if (
+      coercedClientBuild !== null &&
+      coercedServerBuild !== null &&
+      coercedClientBuild.raw !== coercedServerBuild.raw &&
+      coercedClientBuild.raw !== LOCAL_DEV_VERSION
+    ) {
+      const displayVersion = coercedServerBuild.raw;
+      const description = {
+        id: i18nMark(
+          "DC/OS UI has been updated to {version}. Reload the UI for the updated version."
+        ),
+        values: {
+          version: displayVersion
+        }
+      };
+      const notification = new ToastNotification(i18nMark("New UI Available"), {
+        appearance: ToastAppearance.Success,
+        description,
+        primaryActionText: i18nMark("Reload"),
+        callback: updateAvailableCallback
+      });
+      displayNotification(notificationService, notification);
+    }
+  });
+}
+
+function setupUpdateFailedNotification(
+  notificationService: NotificationService
+) {
+  getAction$()
+    .pipe(
+      filter(
+        action =>
+          action.action === UIActions.Error &&
+          action.type === UIActionType.Update
+      )
+    )
+    .subscribe(action => {
+      let notification: ToastNotification;
+      const title = i18nMark("UI Upgrade Failure");
+      const description = i18nMark(
+        "The UI upgrade has failed due to an error."
+      );
+      if (action.value.data !== undefined) {
+        const upgradeVersion: string = action.value.data;
+        const tryAgainCallback = (
+          callbackType: ToastCallbackType,
+          toastNotification: ToastNotification
+        ) => {
+          updateFailedCallback(callbackType, toastNotification, upgradeVersion);
+        };
+        notification = new ToastNotification(title, {
+          appearance: ToastAppearance.Danger,
+          description,
+          primaryActionText: i18nMark("Try Again"),
+          callback: tryAgainCallback
+        });
+      } else {
+        notification = new ToastNotification(title, {
+          appearance: ToastAppearance.Danger,
+          description,
+          callback: dismissNotificationCallback
+        });
+      }
+
+      displayNotification(notificationService, notification);
+    });
+}
+
+function setupRollbackFailedNotification(
+  notificationService: NotificationService
+) {
+  getAction$()
+    .pipe(
+      filter(
+        action =>
+          action.action === UIActions.Error &&
+          action.type === UIActionType.Reset
+      )
+    )
+    .subscribe(() => {
+      const notification = new ToastNotification(
```

Comment:
we could map the stream to the stream of `new ToastNotification` objects then filter the stream by `notification.id`  and that could go directly to the NotificationService, so that we don't need `displayNotification` with extra filtering logic :)

Reply:
I'm coming from seeing streams as flow of data through operators ideally with no extra steps/boilerplate on the end. It's like operating on an array to me `[...actions].filter().map().filter().reduce()`